var/spack/repos/builtin/packages/netcdf-c/package.py

```diff
@@ -46,68 +54,104 @@ class NetcdfC(CMakePackage, AutotoolsPackage):
+    )
 
     variant("mpi", default=True, description="Enable parallel I/O for netcdf-4")
     variant("parallel-netcdf", default=False, description="Enable parallel I/O for classic files")
     variant("hdf4", default=False, description="Enable HDF4 support")
     variant("pic", default=True, description="Produce position-independent code (for shared libs)")
     variant("shared", default=True, description="Enable shared library")
     variant("dap", default=False, description="Enable DAP support")
+    variant("byterange", default=False, description="Enable byte-range I/O")
     variant("jna", default=False, description="Enable JNA support")
     variant("fsync", default=False, description="Enable fsync support")
-    variant("zstd", default=True, description="Enable ZStandard compression", when="@4.9.0:")
+    variant("nczarr_zip", default=False, description="Enable NCZarr zipfile format storage")
     variant("optimize", default=True, description="Enable -O2 for a more optimized lib")
 
-    # It's unclear if cdmremote can be enabled if '--enable-netcdf-4' is passed
-    # to the configure script. Since netcdf-4 support is mandatory we comment
-    # this variant out.
-    # variant('cdmremote', default=False,
-    #         description='Enable CDM Remote support')
+    variant("szip", default=True, description="Enable Szip compression plugin")
+    variant("blosc", default=True, description="Enable Blosc compression plugin")
+    variant("zstd", default=True, description="Enable Zstandard compression plugin")
+
+    # Starting version 4.9.1, nczarr_test/CMakeLists.txt relies on the FILE_PERMISSIONS feature of
+    # the configure_file command, which is only available starting CMake 3.20:
+    depends_on("cmake@3.20:", type="test", when="@4.9.1: build_system=cmake")
 
-    # The patch for 4.7.0 touches configure.ac. See force_autoreconf below.
     with when("build_system=autotools"):
-        depends_on("autoconf", type="build", when="@4.7.0,main")
-        depends_on("automake", type="build", when="@4.7.0,main")
-        depends_on("libtool", type="build", when="@4.7.0,main")
-    # CMake system can use m4, but Windows does not yet support
-    depends_on("m4", type="build", when=sys.platform != "win32")
+        for __s in itertools.chain(["@main"], _force_autoreconf_when):
+            with when(__s):
+                depends_on("autoconf", type="build")
+                depends_on("automake", type="build")
+                depends_on("libtool", type="build")
+                depends_on("m4", type="build")
+        del __s
+
+    # M4 is also needed for the source and man file generation. All the generated source files are
+    # included in the release tarballs starting at least the oldest supported version:
+    depends_on("m4", type="build", when="@main")
+
+    # The man files are included in the release tarballs starting version 4.5.0 but they are not
+    # needed for the Windows platform.
+    with when(sys.platform != "win32"):
```

Comment:
This should be `with when("platform=windows")` as we should be referring to the spec target platform.

Reply:
@johnwparent ok, you have already corrected your message. I don't mind this suggestion. I just took this condition from [your code](https://github.com/spack/spack/blob/385834bd43e1958667b7f75074e07f412d6f58ab/var/spack/repos/builtin/packages/netcdf-c/package.py#L103) and could not find anything similar to `spack.compilers.supported_compilers` (which we use in similar cases), but for the platforms. I now see that this pattern is used in many packages, thank you.

Since you are here, maybe you can comment regarding the [`4.8.1-win-hdf5-with-zlib.patch`](https://github.com/spack/spack/pull/36485/files#diff-9bb18f61f441605ae314f7ccaa7feebb323fe12be6df13a0364f93903977477cR57-R60)?

Reply:
Yes sorry, I autopiloted to hit `comment` before I was done actually writing my comment. My code was also not ideal/correct and I should have caught that in review.
IIRC there's something in the works for the platform equivalent of `spack.compilers.supported_compilers` but is not yet ready for deployment/is still preliminary. The pattern is unwieldy and verbose but is the convention at the moment. 

In general, if you're modifying something that impacts Windows support, absolutely feel free to summon me if you feel the need/are unsure as there is very minimal CI protecting the Windows side of things.

The patch was necessary for me to get this package building against hdf5 with version 4.8.1. IIRC there was an issue raised either on the netcdf-c or hdf5 repos that covered this exact problem and is still open. If you'd like I can try to track that down.

Reply:
Yes, I would like to know more about the patch. As far as I understand, `H5public.h` includes `H5pubconf.h` and the expected `H5_HAVE_ZLIB_H` macro is supposed to be set anyway. I am also curious whether the patch is needed for later versions of NetCDF, which also include `H5public.h`.